File: src/main/java/editor/serialization/AttributeAdapter.java
package editor.serialization;

import java.lang.reflect.Type;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

import editor.database.attributes.CardAttribute;

/**
 * A JSON serializer and deserializer for {@link CardAttribute}.  Basically
 * just uses its {@link CardAttribute#toString()} result for readability.
 * 
 * @author Alec Roelke
 */
public class AttributeAdapter implements JsonSerializer<CardAttribute>, JsonDeserializer<CardAttribute>
{
    @Override
    public CardAttribute deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException
    {
        try
        {
            return CardAttribute.fromString(json.getAsString());
        }
        catch (IllegalArgumentException e)
        {
            throw new JsonParseException(e);
        }
    }

    @Override
    public JsonElement serialize(CardAttribute src, Type typeOfSrc, JsonSerializationContext context)
    {
        return new JsonPrimitive(src.toString());
    }
}

File: src/main/scala/editor/serialization/AttributeAdapter.scala
package editor.serialization

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import com.google.gson.JsonPrimitive
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import editor.database.attributes.CardAttribute
import org.json4s.CustomSerializer
import org.json4s.JString

import java.lang.reflect.Type

/**
 * JSON serializer/deserializer for [[CardAdapter]], which uses its [[CardAdapter.toString]] method for conversion.
 * @author Alec Roelke
 */
class AttributeAdapter extends CustomSerializer[CardAttribute[?, ?]](format => (
  { case JString(attribute) => CardAttribute.parse(attribute).getOrElse(throw IllegalArgumentException(attribute)) },
  { case attribute: CardAttribute[?, ?] => JString(attribute.toString) }
)) with JsonSerializer[CardAttribute[?, ?]] with JsonDeserializer[CardAttribute[?, ?]] {
  override def deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext) = CardAttribute.parse(json.getAsString).getOrElse(throw JsonParseException(json.getAsString))

  override def serialize(src: CardAttribute[?, ?], typeOfSrc: Type, context: JsonSerializationContext) = JsonPrimitive(src.toString)
}